File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) 
    {
        vector<int> dict(256, -1);
        int maxlen = 0, start = 0, n = s.size();
        
        for(int i=0; i<n; i++)
        {
            if(dict[s[i]] != -1)
                start = max(start, dict[s[i]]+1);
            dict[s[i]] = i;
            maxlen = max(maxlen, i-start+1);
        }
        return maxlen;
    }
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        last_seen = {}
        start = longest = 0
        
        for i, ch in enumerate(s):
            if ch in last_seen and last_seen[ch] >= start:
                start = last_seen[ch] + 1
            else:
                longest = max(longest, i - start + 1)
            
            last_seen[ch] = i
        return longest